[MRESOLVER-7] Download dependency POMs in parallel

hwellmann opened the pull request.

michael-o commented:
This commit is massive. I have serious problems understand what you have actually changed because I have never worked on Aether before. There are some noise changes like reformatting, externalizing. etc.  Can you move them into a separate commit? It would make it easier for me to understand your real changes. Can you also provide some explanation why so much change is necessary? Did you already run your changes through Core IT Suite?

hwellmann commented:
Yes, I would have liked it to be smaller, but it's difficult to go back now. The original patch is almost a year old, and I just tried to rework everything to include the changes from master. This was a manual process and not a merge.

The basic idea of my change is quite straightforward: 

* Split loading and processing dependency POMs into two separate steps.
* Load the POMs of the direct dependencies asynchronously and in parallel.
* Wait for these POMs to be loaded and then process them.

This is implemented here:
https://github.com/apache/maven-resolver/pull/2/commits/53ef352e85f1e358367236908b44d3b468689f9c#diff-e8cd0dcf95330cc5029d1cc020c4fc27R378

Everything else is just auxiliary refactoring to make this possible.

`DefaultDependencyCollector` was just so huge and with so many code smells that I had to move things out of the way to really understand what was going on in the class.

I replaced long parameter lists by context objects and copied these objects rather than overwriting them to make the parallel recursion work.

I used the official Eclipse formatter, but it seems it wasn't consistently applied in the past, so this explains some extra whitespace changes.

Regarding the integration test suite, I did run it successfully with my original patch last year. At the moment, it seems to be broken even without my changes, or maybe I didn't invoke it in the correct way. I couldn't find any instructions except https://github.com/apache/maven-integration-testing/blob/master/run-its.sh which doesn't really look up to date.

Maybe we can find more reviewers on the maven-dev list? When I posted the original patch, I had a private message from Jason saying the PR looked ok and he'd like to have it included after the aether-to-apache move, which has now taken so long, unfortunately...

michael-o commented:
Virtually everything you wrote sounds reasonable to me. I still would like someone else to review it as I am not really firm with the code. Please raise on the dev mailing list.

michael-o commented:
Don't forget about the mailing list review if you want this to be merged.

ChristianSchulte commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDependencyCollector.java:
```diff
@@ -89,6 +94,8 @@
 
     private VersionRangeResolver versionRangeResolver;
 
+    private ExecutorService executor = Executors.newFixedThreadPool( 5 );
```
You want this to scale so limiting this to 5 should not be done. In Java 8 there is a [commonPool](http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html#commonPool--) sized by the number of available processors. Is there a way to auto size that based on some criteria? Number of available processor? Bandwidth available?

ChristianSchulte commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDependencyCollector.java:
```diff
@@ -351,157 +375,215 @@ private static String getId( Artifact a )
-                                    DependencyTraverser depTraverser, VersionFilter verFilter, Dependency dependency )
+    private Future<DependencyContext> asyncProcessDependency( final DependencyContext dc )
     {
+        return executor.submit( new Callable<DependencyContext>()
```
Are you sure this can be done? When resolving a child node, that node can change the repositories in use. If all child nodes will be processed asynchronously, that may not work anymore.

ChristianSchulte commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDependencyCollector.java:
```diff
@@ -351,157 +375,215 @@ private static String getId( Artifact a )
-
-            rangeResult = cachedResolveRangeResult( rangeRequest, args.pool, args.session );
+            VersionRangeRequest rangeRequest = createVersionRangeRequest( args, context.getRepositories(), dependency );
+            VersionRangeResult rangeResult = cachedResolveRangeResult( rangeRequest, args.pool, args.session );
```
If this will download files to the local repository, this cannot be done asynchronously as the local repository is not thread-safe.

ChristianSchulte commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDependencyCollector.java:
```diff
@@ -351,157 +375,215 @@ private static String getId( Artifact a )
+                dc.originalArtifact = originalArtifact;
+                dc.managedDependency = d;
+                dc.futureDescriptorResult =
+                    getArtifactDescriptorResult( args, results, noDescriptor, d, descriptorRequest );
```
Here as well. It will modify the local repository asynchronously.

ChristianSchulte commented on maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDependencyCollector.java:
```diff
@@ -351,157 +375,215 @@ private static String getId( Artifact a )
-            args.pool.toKey( d.getArtifact(), childRepos, childSelector, childManager, childTraverser, childFilter );
+        Object key = args.pool.toKey( d.getArtifact(), childContext );
 
         List<DependencyNode> children = args.pool.getChildren( key );
```
This will not work when fetched asynchronously. It's an implementation side effect I do not like myself. Child node lists are cached and re-used - no matter what parent node has lead to the resolution of the child node list. It's not easy to remove that caching. It is used for processing cycles as well.

ChristianSchulte commented:
Would it be possible to split this into multiple smaller commits so that reviewing becomes easier? Changing behaviour of the 'DefaultDepencencyCollector' by accident is a recipe for disaster, for example. You have extracted code to various methods. Maybe a commit for each new method? So that the last commit is about parallelization only and not about refactoring code? I would not mind having to review 20+ commits for this when the first 19 commits do not change any behaviour but only are about preparing/refactoring code for parallelization.

hapylestat commented:
does this have chance to be completed?    maven still works as piece of shit with downloading dependencies

michael-o commented:
> does this have chance to be completed? maven still works as piece of shit with downloading dependencies

If someone rebases on top of master, yes. This won't speed up artifact download, POMs only.

mosabua commented:
While this is a nice feature to have I do not think we can reasonably merge this PR as it stands and I suggest to close it. I dont think anyone can tell what actually changes and hence the risk of applying this is way too high.

slachiewicz commented:
I have prepared new version - please check.

krichter722 commented:
> While this is a nice feature to have I do not think we can reasonably merge this PR as it stands and I suggest to close it. I dont think anyone can tell what actually changes and hence the risk of applying this is way too high.

@mosabua That sounds odd. Shouldn't you always quickly know whether a change works or not by applying the software's unit and integration tests on it?

mosabua commented:
@krichter722 while it sounds odd it is the reality. Test coverage within the library is just not big enough to absorb such a large changeset and even beyond that nobody can read it and understand what was changed. To really test resolver you have to also build it and then use it in a custom build of Maven itself with the new resolver and run the Maven IT test. Thats a largish undertaking and it is just easier when changes are more contained. So yes.. in an ideal world you are right, in reality things are a bit more complicated.

michael-o commented:
> To really test resolver you have to also build it and then use it in a custom build of Maven itself with the new resolver and run the Maven IT test. Thats a largish undertaking and it is just easier when changes are more contained.

This is actually what I do...with PRs

gdubicki commented:
I propose to close this PR in favor of https://github.com/apache/maven-resolver/pull/10 .

michael-o commented:
@gdubicki I will close this implicitly with the merge into master to maintain the linked information.

michael-o commented:
Closing this, ti has been superseded.

jira-importer commented:
Resolve #817

jira-importer commented:
Resolve #892

jira-importer commented:
Resolve #817

jira-importer commented:
Resolve #892
